src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs

```diff
@@ -182,29 +185,43 @@ private async Task CreateConstantFieldActionsAsync(State state, ArrayBuilder<Cod
                 // If something is a constant, and it doesn't access any other locals constants,
                 // then we prefer to offer to generate a constant field instead of a constant
                 // local.
-                var action1 = CreateAction(state, allOccurrences: false, isConstant: true, isLocal: false, isQueryLocal: false);
-                if (await CanGenerateIntoContainerAsync(state, action1, cancellationToken).ConfigureAwait(false))
-                {
-                    actions.Add(action1);
-                }
-
-                var action2 = CreateAction(state, allOccurrences: true, isConstant: true, isLocal: false, isQueryLocal: false);
-                if (await CanGenerateIntoContainerAsync(state, action2, cancellationToken).ConfigureAwait(false))
+                if (CanGenerateIntoContainer(state, cancellationToken))
                 {
-                    actions.Add(action2);
+                    actions.Add(CreateAction(state, allOccurrences: false, isConstant: true, isLocal: false, isQueryLocal: false));
+                    actions.Add(CreateAction(state, allOccurrences: true, isConstant: true, isLocal: false, isQueryLocal: false));
                 }
             }
         }
 
-        private async Task<bool> CanGenerateIntoContainerAsync(State state, CodeAction action, CancellationToken cancellationToken)
+        protected int GetFieldInsertionIndex(
+            bool isConstant, TTypeDeclarationSyntax oldType, TTypeDeclarationSyntax newType, CancellationToken cancellationToken)
         {
-            var result = await IntroduceFieldAsync(
-                state.Document, state.Expression,
-                allOccurrences: false, isConstant: state.IsConstant, cancellationToken: cancellationToken).ConfigureAwait(false);
+            var preferredInsertionIndex = isConstant
+                ? DetermineConstantInsertPosition(oldType, newType)
+                : DetermineFieldInsertPosition(oldType, newType);
+
+            var legalInsertionIndices = this.GetInsertionIndices(oldType, cancellationToken);
+            if (legalInsertionIndices[preferredInsertionIndex])
+            {
+                return preferredInsertionIndex;
+            }
+
+            // location we wanted to insert into isn't legal (i.e. it's hidden).  Try to find a
+            // non-hidden location.
+            var legalIndex = legalInsertionIndices.IndexOf(true);
+            if (legalIndex >= 0)
+            {
+                return legalIndex;
+            }
 
-            var destination = result.Item2;
-            var insertionIndex = result.Item3;
+            // Couldn't find a viable non-hidden position.  Fall back to the computed position we
+            // wanted originally.
+            return preferredInsertionIndex;
```

Comment:
This doesn't seem to be reachable since we made sure in `CanGenerateIntoContainer` there's must be at least one legal position, right?

Reply:
not quite.  We only check CanGenerateIntoContainer when *generating a local that we think might be better as field*.   i.e.:

```c#
                // If something is a constant, and it doesn't access any other locals constants,
                // then we prefer to offer to generate a constant field instead of a constant
                // local.
```

We don't do this for the normal 'generate field/constant' codepth.  So we can get into this code without having checked that.  note: this is behavior that existed before for the non-local case.

Reply:
Answered.  `IntroduceFieldAsync` has multiple people that call it.